File: solutions/problems/arraysStrings/validAnagram.ts
/**
 * Approach: compare character-frequency counts.
 * Time: O(n). Space: O(1) (bounded alphabet).
 */

export function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) return false;
  const counts = new Map<string, number>();
  for (const ch of s) counts.set(ch, (counts.get(ch) ?? 0) + 1);
  for (const ch of t) {
    const remaining = (counts.get(ch) ?? 0) - 1;
    if (remaining < 0) return false;
    counts.set(ch, remaining);
  }
  return true;
}

File: problems/arrays_strings/valid_anagram.py
"""
LeetCode 242. Valid Anagram

Return True if `t` is an anagram of `s` (same letters, same counts).

Approach: compare character-frequency counts. Equal-length strings are
anagrams iff their Counters match.

Time:  O(n)
Space: O(1) (bounded alphabet)
"""

from collections import Counter


def is_anagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False
    return Counter(s) == Counter(t)


if __name__ == "__main__":
    assert is_anagram("anagram", "nagaram") is True
    assert is_anagram("rat", "car") is False
    assert is_anagram("", "") is True
    print("All tests passed.")
